File: sdk/go/idempotency.go
package faas

import (
	"context"
	"net/http"
)

// IdempotencyKey is the typed string callers pass to WithIdempotencyKey
// to pin a stable key for retry-safe mutating calls. Any non-empty
// string is accepted; the server's replay middleware (apid/server.go
// ::idempotent) caches the response for 24h keyed on this value.
//
// IdempotencyKey has no validation in the SDK; the server validates
// the format. Callers that want a UUIDv4 per attempt should use
// google/uuid (or crypto/rand) and pass the result here.
type IdempotencyKey string

type idempotencyKeyCtxKey struct{}

// WithIdempotencyKey returns a context.Context that carries the key
// for the next HTTP request made through this Client. The key is
// extracted by the idempotency round-tripper (installed by NewClient)
// and injected as the Idempotency-Key request header before the
// auto-mint in internal/api/client.do fires.
//
// Pass the returned context to any mutating Client method:
//
//	ctx = faas.WithIdempotencyKey(ctx, "deploy-attempt-3")
//	dep, err := c.Deploy(ctx, slug, req)
//
// If the caller does not supply a key, the SDK auto-mints a UUIDv4
// per call. Both paths are safe; the difference is that supplying
// a stable key makes retries deterministic.
func WithIdempotencyKey(ctx context.Context, key IdempotencyKey) context.Context {
	if key == "" {
		return ctx
	}
	return context.WithValue(ctx, idempotencyKeyCtxKey{}, key)
}

// idempotencyKeyFromContext extracts the key. Returns "" if absent.
func idempotencyKeyFromContext(ctx context.Context) IdempotencyKey {
	v, _ := ctx.Value(idempotencyKeyCtxKey{}).(IdempotencyKey)
	return v
}

// idempotencyRoundTripper injects the Idempotency-Key header from the
// request context before delegating to the wrapped transport. When
// the context carries no key, the shim is a no-op and the auto-mint
// in internal/api/client.do takes over.
type idempotencyRoundTripper struct {
	inner http.RoundTripper
}

func newIdempotencyRoundTripper(inner http.RoundTripper) *idempotencyRoundTripper {
	if inner == nil {
		inner = http.DefaultTransport
	}
	return &idempotencyRoundTripper{inner: inner}
}

func (r *idempotencyRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
	if key := idempotencyKeyFromContext(req.Context()); key != "" {
		// An explicit caller-supplied key from WithIdempotencyKey
		// always wins — even over the auto-mint that
		// internal/api/client.do may have already set. This matches
		// the documented "stable key for retries" contract: if the
		// caller pinned a key, the auto-mint is a no-op.
		//
		// Clone the header so we don't mutate the caller's request
		// (http.Request.Header is shared via the Transport contract).
		req2 := req.Clone(req.Context())
		req2.Header = req.Header.Clone()
		req2.Header.Set("Idempotency-Key", string(key))
		req = req2
	}
	return r.inner.RoundTrip(req)
}

File: sdk/node/src/idempotency.ts
// src/idempotency.ts — request-scoped Idempotency-Key support.
//
// The contract mirrors `sdk/go/idempotency.go` and
// `pkg/api/internal/api/client.go::do`:
//   * Every mutating call (POST/PUT/PATCH/DELETE) MUST carry an
//     `Idempotency-Key` header. The server replays the same response
//     for the same key within 24h (apid's idempotency middleware).
//   * GET/HEAD skip the header — the server doesn't dedupe reads.
//   * Default behaviour: the SDK auto-mints a UUIDv4-shaped key on
//     every mutating call.
//   * Opt-in: callers can pin a stable key via `withIdempotencyKey`
//     for retry semantics that survive across processes (CI deploys,
//     idempotent batch jobs, etc.).
//
// The async-local-storage wire-in is a placeholder for a PR 11
// follow-up if the docs team requests it; today the explicit
// `OpenAPI.HEADERS['Idempotency-Key']` default is sufficient.

import { randomUUID } from 'node:crypto';

/** Type alias for an idempotency key. Any non-empty string ≤ 255 chars
 *  is accepted; the server doesn't validate the format. */
export type IdempotencyKey = string;

/**
 * Default key generator. UUIDv4 is canonical because the server's
 * 24h replay window keys on string equality — collisions across the
 * window are vanishingly improbable (UUIDv4 = 122 random bits).
 *
 * Re-exported as a named function so callers can plug in their own
 * (e.g. for tests that need a deterministic key).
 */
export function mintIdempotencyKey(): IdempotencyKey {
  return randomUUID();
}

/**
 * Mutating methods that REQUIRE an Idempotency-Key. Listed explicitly
 * rather than computed from the method because we want a single
 * source of truth for the GET-skip rule.
 */
export const MUTATING_METHODS = new Set([
  'POST',
  'PUT',
  'PATCH',
  'DELETE',
]);

/** Pure predicate: does this HTTP method require an Idempotency-Key? */
export function isMutating(method: string): boolean {
  return MUTATING_METHODS.has(method.toUpperCase());
}

/**
 * Placeholder for a future AsyncLocalStorage-based key wire-in.
 * Today the wrapper reads `OpenAPI.HEADERS['Idempotency-Key']`
 * (which `FaaSClient` sets at construction time) so per-call keys
 * require rebuilding the client. This function exists so a future
 * PR can swap the impl without changing call sites:
 *
 *     const key = withIdempotencyKey(ctx, 'stable-key');
 *     await streamSse(resp, signal); // future: reads key from ctx
 *
 * Returns the key unchanged today; PR 11 may upgrade it to a
 * `AsyncResource<IdempotencyKey>` wrapper.
 */
export function withIdempotencyKey(_ctx: unknown, key: IdempotencyKey): IdempotencyKey {
  return key;
}